### PR TITLE
Add ability to serch books

### PR DESCRIPTION
**Pull request by petro2061**
Search for books using certain parameters is implemented: titles, authors, isbnes, prices. Implementation is based on Creteria API and Spesification. For this, the following interfaces have been added: SpecificationBuilde, SpecificationProvider, SpecificationProviderManager and their corresponding implementation.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/repository/book/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> specificationProviderList;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        return specificationProviderList
+                .stream()
+                .filter(provider -> provider.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Can't find correct "
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/repository/book/BookSpecificationProviderManager.java]: throw more suitable NoSuchElementException

File: src/main/java/project/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    @Override
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) ->
+                root.get("author").in(Arrays.stream(params).toList());
+    }
+
+    @Override
+    public String getKey() {
+        return "author";
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: extract repeated literal to a constant

File: src/main/java/project/onlinebookstore/dto/BookSearchParameters.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package project.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+
+public record BookSearchParameters(String[] titles,
+                                   String[] authors,
+                                   String[] isbnes,
```
Comment (okuzan) [inline src/main/java/project/onlinebookstore/dto/BookSearchParameters.java]:
```suggestion
                                   String[] isbns,
```

**Review by okuzan — CHANGES_REQUESTED** [review]

**Comment by petro2061** [comment]
Fixed field name and extracted repeated literals to a constant.

**Comment by petro2061** [comment]
Fixed field name and extracted repeated literals to a constant.

File: src/main/java/project/onlinebookstore/repository/book/BookSpecificationProviderManager.java (hunk 1 above)
Comment (Elena-Bruyako) [reply src/main/java/project/onlinebookstore/repository/book/BookSpecificationProviderManager.java]: not fixed

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Please, pay attention to the mentor comment about changes are not related to the HW.

**Comment by petro2061** [comment]
Added correct class exception to getSpecificationProvider method

**Review by okuzan — APPROVED** [review]
Close PR, old branch should be deleted, Merge into master, create new branch out of master for a new HW

**Pull request merged**